File: cpp/example_onnx.cpp
#include "helper.h"
#include <iostream>
#include <filesystem>
#include <algorithm>
#include <string>
#include <vector>

namespace fs = std::filesystem;

struct Args {
    std::string onnx_dir = "../assets/onnx";
    int total_step = 5;
    int n_test = 4;
    std::vector<std::string> voice_style = {"../assets/voice_styles/M1.json"};
    std::vector<std::string> text = {
        "This morning, I took a walk in the park, and the sound of the birds and the breeze was so pleasant that I stopped for a long time just to listen."
    };
    std::string save_dir = "results";
};

auto splitString = [](const std::string& str, char delim) {
    std::vector<std::string> result;
    size_t start = 0, pos;
    while ((pos = str.find(delim, start)) != std::string::npos) {
        result.push_back(str.substr(start, pos - start));
        start = pos + 1;
    }
    result.push_back(str.substr(start));
    return result;
};

Args parseArgs(int argc, char* argv[]) {
    Args args;
    for (int i = 1; i < argc; i++) {
        std::string arg = argv[i];
        if (arg == "--onnx-dir" && i + 1 < argc) args.onnx_dir = argv[++i];
        else if (arg == "--total-step" && i + 1 < argc) args.total_step = std::stoi(argv[++i]);
        else if (arg == "--n-test" && i + 1 < argc) args.n_test = std::stoi(argv[++i]);
        else if (arg == "--voice-style" && i + 1 < argc) args.voice_style = splitString(argv[++i], ',');
        else if (arg == "--text" && i + 1 < argc) args.text = splitString(argv[++i], '|');
        else if (arg == "--save-dir" && i + 1 < argc) args.save_dir = argv[++i];
    }
    return args;
}

int main(int argc, char* argv[]) {
    std::cout << "=== TTS Inference with ONNX Runtime (C++) ===\n\n";
    
    // --- 1. Parse arguments --- //
    Args args = parseArgs(argc, argv);
    int total_step = args.total_step;
    int n_test = args.n_test;
    std::string save_dir = args.save_dir;
    std::vector<std::string> voice_style_paths = args.voice_style;
    std::vector<std::string> text_list = args.text;
    
    if (voice_style_paths.size() != text_list.size()) {
        std::cerr << "Error: Number of voice styles (" << voice_style_paths.size() 
                  << ") must match number of texts (" << text_list.size() << ")\n";
        return 1;
    }
    
    int bsz = voice_style_paths.size();
    
    // --- 2. Load Text to Speech --- //
    Ort::Env env(ORT_LOGGING_LEVEL_WARNING, "TTS");
    Ort::MemoryInfo memory_info = Ort::MemoryInfo::CreateCpu(
        OrtAllocatorType::OrtArenaAllocator, OrtMemType::OrtMemTypeDefault
    );
    
    auto text_to_speech = loadTextToSpeech(env, args.onnx_dir, false);
    std::cout << std::endl;
    
    // --- 3. Load Voice Style --- //
    auto style = loadVoiceStyle(voice_style_paths, true);
    
    // --- 4. Synthesize speech --- //
    fs::create_directories(save_dir);
    
    for (int n = 0; n < n_test; n++) {
        std::cout << "\n[" << (n + 1) << "/" << n_test << "] Starting synthesis...\n";
        
        auto result = timer("Generating speech from text", [&]() {
            return text_to_speech->call(memory_info, text_list, style, total_step);
        });
        
        int sample_rate = text_to_speech->getSampleRate();
        int wav_shape_1 = result.wav.size() / bsz;
        
        for (int b = 0; b < bsz; b++) {
            std::string fname = sanitizeFilename(text_list[b], 20) + "_" + std::to_string(n + 1) + ".wav";
            int wav_len = static_cast<int>(sample_rate * result.duration[b]);
            
            std::vector<float> wav_out(
                result.wav.begin() + b * wav_shape_1,
                result.wav.begin() + b * wav_shape_1 + wav_len
            );
            
            std::string output_path = save_dir + "/" + fname;
            writeWavFile(output_path, wav_out, sample_rate);
            std::cout << "Saved: " << output_path << "\n";
        }
        
        clearTensorBuffers();
    }
    
    std::cout << "\n=== Synthesis completed successfully! ===\n";
    return 0;
}

File: swift/Sources/ExampleONNX.swift
import Foundation
import OnnxRuntimeBindings

struct Args {
    var useGpu: Bool = false
    var onnxDir: String = "assets/onnx"
    var totalStep: Int = 5
    var nTest: Int = 4
    var voiceStyle: [String] = ["assets/voice_styles/M1.json"]
    var text: [String] = ["This morning, I took a walk in the park, and the sound of the birds and the breeze was so pleasant that I stopped for a long time just to listen."]
    var saveDir: String = "results"
}

func parseArgs() -> Args {
    var args = Args()
    let arguments = CommandLine.arguments
    
    var i = 1
    while i < arguments.count {
        let arg = arguments[i]
        
        switch arg {
        case "--use-gpu":
            args.useGpu = true
        case "--onnx-dir":
            if i + 1 < arguments.count {
                args.onnxDir = arguments[i + 1]
                i += 1
            }
        case "--total-step":
            if i + 1 < arguments.count {
                args.totalStep = Int(arguments[i + 1]) ?? 5
                i += 1
            }
        case "--n-test":
            if i + 1 < arguments.count {
                args.nTest = Int(arguments[i + 1]) ?? 4
                i += 1
            }
        case "--voice-style":
            if i + 1 < arguments.count {
                args.voiceStyle = arguments[i + 1].components(separatedBy: ",")
                i += 1
            }
        case "--text":
            if i + 1 < arguments.count {
                args.text = arguments[i + 1].components(separatedBy: "|")
                i += 1
            }
        case "--save-dir":
            if i + 1 < arguments.count {
                args.saveDir = arguments[i + 1]
                i += 1
            }
        default:
            break
        }
        
        i += 1
    }
    
    return args
}

@main
struct ExampleONNX {
    static func main() async {
        print("=== TTS Inference with ONNX Runtime (Swift) ===\n")
        
        // --- 1. Parse arguments --- //
        let args = parseArgs()
        
        guard args.voiceStyle.count == args.text.count else {
            print("Error: Number of voice styles (\(args.voiceStyle.count)) must match number of texts (\(args.text.count))")
            return
        }
        
        let bsz = args.voiceStyle.count
        
        do {
            let env = try ORTEnv(loggingLevel: .warning)
            
            // --- 2. Load TTS components --- //
            let textToSpeech = try loadTextToSpeech(args.onnxDir, args.useGpu, env)
            
            // --- 3. Load voice styles --- //
            let style = try loadVoiceStyle(args.voiceStyle, verbose: true)
            
            // --- 4. Synthesize speech --- //
            try? FileManager.default.createDirectory(atPath: args.saveDir, withIntermediateDirectories: true)
            
            for n in 0..<args.nTest {
                print("\n[\(n + 1)/\(args.nTest)] Starting synthesis...")
                
                let (wav, duration) = try timer("Generating speech from text") {
                    try textToSpeech.call(args.text, style, args.totalStep)
                }
                
                // Save outputs
                let wavLen = wav.count / bsz
                for i in 0..<bsz {
                    let fname = "\(sanitizeFilename(args.text[i], maxLen: 20))_\(n + 1).wav"
                    let actualLen = Int(Float(textToSpeech.sampleRate) * duration[i])
                    
                    let wavStart = i * wavLen
                    let wavEnd = min(wavStart + actualLen, wavStart + wavLen)
                    let wavOut = Array(wav[wavStart..<wavEnd])
                    
                    let outputPath = "\(args.saveDir)/\(fname)"
                    try writeWavFile(outputPath, wavOut, textToSpeech.sampleRate)
                    print("Saved: \(outputPath)")
                }
            }
            
            print("\n=== Synthesis completed successfully! ===")
            
        } catch {
            print("Error during inference: \(error)")
            exit(1)
        }
    }
}
